pyccel/parser/semantic.py

```diff
@@ -3682,13 +3709,36 @@ def _visit_ClassDef(self, expr):
         for i in methods:
             self._visit_FunctionDef(i)
 
+        if not any(method.name == '__del__' for method in methods):
+            argument = FunctionDefArgument(Variable(cls.name, 'self', cls_base = cls))
+            scope = self.create_new_function_scope('__del__')
+            del_method = FunctionDef('__del__', [argument], (), [Pass()], cls_name=cls.name, scope=scope)
+            self.exit_function_scope()
+            self.insert_function(del_method)
+            cls.add_new_method(del_method)
+
+        if not self._allocs:
+            self._allocs.append([])
```

Comment:
Can you explain this line? Why do we not systematically add a scope for the `ClassDef` variables? Won't this implementation potentially result in the enclosing scope's `_allocs` being popped in `self._garbage_collector`?

E.g. The following will fail:
```python
from numpy import ones

class Point(object):
    def __init__(self : 'Point', n : 'int'):
        self.y = ones(n)

    def __del__(self : 'Point'):
        pass

x = ones(4)
```
please add this as a test and fix this bug